### DISCUSSION TITLE
DataTable performance

### DISCUSSION THREAD

**Discussion by vegegoku** [DominoKit/domino-ui#627] [General]
A discussion aimed at the performance of datatable with large data collection especially while using the LocalListStore.

**Comment by FrankHossfeld** [comment]
How do you define large data collections ? ;-)

**Reply by vegegoku** [reply]
I did a quick test with a very simple HTML and Javascript and it renders 5000 record rows super fast 

```html
<html>
<body>
<table>
    <thead>
    <tr>
        <th>name</th>
        <th>code</th>
    </tr>
    </thead>
    <tbody id="s-tbody">

    </tbody>
</table>

<script type="application/javascript">
    let tbody = document.getElementById("s-tbody");
    for (let i = 0; i< 5000; i++){
        let tr = document.createElement("tr");
        tbody.appendChild(tr)
        let name = document.createElement("td");
        name.textContent = "name - "+i;
        let code = document.createElement("td");
        code.textContent = "code - "+i;
        tr.appendChild(name);
        tr.appendChild(code);
    }
</script>
</body>
</html>
```

I dont want the exact same performance for 5000 records because we knew in an actual rich datatable it is not that simple, but I want to make it reasonable fast and make sure that we are not just missing on something that is slowing things down.

**Comment by FrankW76** [comment]
Can this also be investigated for Select with many objects.
Seems to be going slow when filling up with 3000 items.

**Comment by vegegoku** [comment]
I just did a quick test 
![500records](https://user-images.githubusercontent.com/5379625/149935717-89f4a292-238b-4737-9354-c28cf4237ffe.gif)

```java
public class App implements EntryPoint {

    /**
     * This is the entry point method.
     */
    public void onModuleLoad() {
        var layout = Layout.create("Domino-ui starter")
                .show(ColorScheme.BLUE);

        List<City> cities = new ArrayList<>();
        for(int i=0; i<5000;i++){
            cities.add(new City("name - "+i, "code - "+i));
        }

        TableConfig<City> tableConfig = new TableConfig<City>()
                .addColumn(ColumnConfig.<City>create("name", "Name")
                        .setCellRenderer(cellInfo -> DominoElement.div().appendChild(TextNode.of(cellInfo.getRecord().getName())).element())
                )
                .addColumn(ColumnConfig.<City>create("code", "code")
                        .setCellRenderer(cellInfo -> DominoElement.div().appendChild(TextNode.of(cellInfo.getRecord().getCode())).element())
                );

        LocalListDataStore<City> localListDataStore = new LocalListDataStore<>();
        DataTable<City> cityDataTable= new DataTable<>(tableConfig, localListDataStore);

        layout.getContentPanel().appendChild(Card.create()
                .appendChild(cityDataTable)
        );

        localListDataStore.setData(cities);
    }
}
```
It pretty fast even though in this example I had to loop the 5000 records twice unlike the javascript sample above.

**Reply by vegegoku** [reply]
Even tested with header filters and it is still fast :thinking: 

@FrankW76 Can you post a sample where this is slow for you.

**Reply by FrankW76** [reply]
In your example, how long does the : localListDataStore.setData(cities); take?

For me that line is about 12 seconds.
After this the table is fast. Except for the filtering. But we use Reflect for this, which may be causing issues.

I could also take a look in the cellRenderers code as we do quite a lot of stuff in those. If something is chocking it up.

**Reply by FrankW76** [reply]
I managed to make it a lot faster with optimizing cellrenders. Like from 12 seconds to 2 or 3 seconds.


One column that is giving me some issues is with boolean columns. If add this column it will directly add 4 or 5 seconds extra to the time to load.

`            
boolean value = Boolean.parseBoolean(String.valueOf(Reflect.get(cell.getTableRow().getRecord(), columnIdentifier)));
SwitchButton switchButton = SwitchButton.create();
switchButton.setColor(Color.BLUE);
switchButton.setValue(value);
if (!enabled) switchButton.disable();
switchButton.getOnOffLabelElement().style().setCursor("default");

if (callback != null)
{
   int iValue = Integer.parseInt(String.valueOf(Reflect.get(cell.getTableRow().getRecord(), "id")));
   switchButton.addChangeHandler(isOn -> callback.onAction(iValue, isOn));
}

return switchButton.element();`


This take some time for each cell.
But when rendering 3000 lines, this offcourse adds up...

**Reply by vegegoku** [reply]
I use an approach for common column configs that does not require reflection by passing some sort of suppliers functions like this :

first define a supplier function
```java

    @FunctionalInterface
    public interface BooleanSupplier<T> {
        boolean getBoolean(T bean);
    }
```

then create a static factory method for the column config : 

```java

    public static <T> ColumnConfig<T> yesNo(BooleanSupplier<T> booleanSupplier, String name, String label) {
        return ColumnConfig.<T>create(name, Labels.get(label))
                .sortable()
                .setCellRenderer(cellInfo ->
                        BooleanTableCell.trueFalseCell(booleanSupplier.getBoolean(cellInfo.getRecord())));
    }
```

then whenever I need a column that display `Yes/No` based on boolean value I add it like this : 

```java
tableConfig
...
                .addColumn(CommonColumns.yesNo(BillerUIResource::getRefundAllowed, "name", "label"))
...

```
And you can even make a factory class for those common column that delegate to the static factory method based on field type.

You can either use static checks to find the field type or do a code generattion .. IMHO this is much better than using something like reflect which is even affect GWT optimization.

**Comment by HAC-jh** [comment]
I also encountered a slow perfomance at the DataTable especially having many columns and rows and using a Select as a cell renderer. This leads to loading times for the table up to several minutes. I will try to provide an example for that

**Reply by vegegoku** [reply]
Have you tried to test with the latest HEAD-SNAPSHOT.

sorry handling this from sickness bed I need few days to recover.

**Reply by HAC-jh** [reply]
I will try it, thx for the hint. Take your time to rest and get well soon
